### PR TITLE
Add zero-copy HTTP/1.x parser with multi-arch SIMD (2-9x faster than llhttp)

### PR DESCRIPTION
**Pull request by hurricane1026**
## Summary

- Hand-written zero-copy HTTP/1.x request parser with SIMD-accelerated scanning
- 6 SIMD backends: scalar, SSE2, AVX2, AVX-512, NEON, SVE (compile-time selected via `-DSIMD_ARCH=`)
- Nanobench-inspired benchmark framework (`testing/bench.h`), zero-stdlib
- 113 tests, ~1400 assertions — corpus ported from llhttp (Node.js) and nginx test suites
- **2-9x faster than llhttp** across all payload sizes, up to 4.1 GB/s throughput

## Performance (SSE2, i7-10700)

| Payload | rue | llhttp | Speedup | Throughput |
|---|---|---|---|---|
| simple_get (18B) | 5 ns | 44 ns | **8.8x** | 3.6 GB/s |
| browser_get (365B) | 107 ns | 267 ns | **2.5x** | 3.4 GB/s |
| api_post (217B) | 82 ns | 231 ns | **2.8x** | 2.6 GB/s |
| curl_get (154B) | 38 ns | 121 ns | **3.2x** | 4.1 GB/s |
| chunked_post (170B) | 61 ns | 232 ns | **3.8x** | 2.8 GB/s |

## Key techniques

- **Lookup-table character classification** (from picohttpparser): 256-byte tables for token, header-value, URI validation
- **SSE2/AVX2/AVX-512 scanning**: `find_header_end`, `scan_header_value`, `scan_uri`, `scan_header_name` — 16/32/64 bytes per iteration
- **Single-pass parsing**: no pre-scan for `\r\n\r\n` on happy path; cold-path disambiguation for Error vs Incomplete
- **Integer method matching**: `load_u32` against `"GET "`, `"POST"` etc. — no scan loop
- **u64 version check**: `"HTTP/1.1"` compared as single 8-byte integer
- **u64-batched `str_ci_eq`**: 8 bytes per iteration for semantic header matching
- **First-byte + length dispatch**: avoids `str_ci_eq` for non-semantic headers

## Files

| Path | Purpose |
|---|---|
| `include/rout/runtime/http_parser.h` | Data structures: HttpMethod, ParsedRequest, HttpParser |
| `src/runtime/http_parser.cc` | Parser core: single-pass, integer tricks, semantic headers |
| `src/runtime/simd/` | 6 SIMD backends (scalar/sse2/avx2/avx512/neon/sve) |
| `testing/bench.h` | Benchmark framework |
| `testing/test.h` | Test framework (moved from include/rout/) |
| `bench/bench_http_parser.cc` | rue vs llhttp benchmark |
| `tests/test_http_parser.cc` | 113 tests, ~1400 assertions |
| `test/fixtures/llhttp/` | Raw llhttp test fixtures (13 markdown files) |
| `third_party/llhttp/` | llhttp v9.2.1 source for benchmark comparison |

## Test plan

- [x] `./dev.sh test` — 113 HTTP parser tests + all existing tests pass
- [x] Benchmark with SSE2, AVX2 backends
- [x] Incremental parsing (byte-by-byte, split-at-every-position)
- [x] Invalid request corpus (40+ malformed requests return Error)
- [ ] Test on ARM (NEON/SVE) — requires aarch64 hardware

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: src/runtime/http_parser.cc [hunk 1]
```diff
@@ -0,0 +1,403 @@
...
+        if (UNLIKELY(colon_pos == name_start)) return ParseStatus::Error;
+        u32 name_len = colon_pos - name_start;
+        pos = colon_pos + 1;
+
+        // Skip OWS — hot path: single space after colon
+        if (UNLIKELY(pos >= len)) goto maybe_incomplete;
+        if (LIKELY(buf[pos] == ' ')) {
+            pos++;
+        }
+        while (UNLIKELY(buf[pos] == ' ' || buf[pos] == '\t')) pos++;
```
[the remark this answers is not in the document]
Comment (hurricane1026) [reply src/runtime/http_parser.cc]: Fixed. Added `pos < len` guard to OWS while loop and `pos >= len → Incomplete` check after.

File: src/runtime/http_parser.cc [hunk 2]
```diff
@@ -0,0 +1,403 @@
...
+        u32 lo = load_u32(val) | 0x20202020U;
+        if (lo == u32_lit('c', 'l', 'o', 's') && (val[4] | 0x20) == 'e') {
+            req->keep_alive = false;
+        }
+    } else if (vlen == 10) {
+        // "keep-alive"
+        u64 v = load_u64(val) | 0x2020202020202020ULL;
+        if (v == u64_lit("keep-ali") && (val[8] | 0x20) == 'v' && (val[9] | 0x20) == 'e') {
+            req->keep_alive = true;
+        }
```
[the remark this answers is not in the document]
Comment (hurricane1026) [reply src/runtime/http_parser.cc]: Fixed. Replaced exact-value matching with comma-separated token list parsing. Handles `close, upgrade` etc.

File: include/rout/runtime/http_parser.h [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+
+    void reset() {
+        parsed_offset = 0;
+        header_end = 0;
+    }
+
+    // Parse request-line + headers from buf[0..len).
+    // On Complete, populates `req` and sets `header_end`.
+    // On Incomplete, preserves `parsed_offset` for next call.
+    ParseStatus parse(const u8* buf, u32 len, ParsedRequest* req);
```
[the remark this answers is not in the document]
Comment (hurricane1026) [reply include/rout/runtime/http_parser.h]: Fixed. Removed `parsed_offset` field entirely — parser always reparses from `buf[0]`.

File: src/runtime/simd/sse2.cc [hunk 4]
```diff
@@ -0,0 +1,122 @@
...
+    const __m128i vcr = _mm_set1_epi8('\r');
+    const __m128i v20 = _mm_set1_epi8(0x20);
+    const __m128i vht = _mm_set1_epi8(0x09);
+    const __m128i vdel = _mm_set1_epi8(0x7F);
+
+    while (pos + 16 <= end) {
+        __m128i chunk = _mm_loadu_si128(reinterpret_cast<const __m128i*>(buf + pos));
+        int cr_mask = _mm_movemask_epi8(_mm_cmpeq_epi8(chunk, vcr));
+        __m128i bad =
+            _mm_or_si128(_mm_andnot_si128(_mm_cmpeq_epi8(chunk, vht), _mm_cmplt_epi8(chunk, v20)),
```
[the remark this answers is not in the document]
Comment (hurricane1026) [reply src/runtime/simd/sse2.cc]: Fixed. Applied xor-0x80 trick: `(chunk ^ 0x80) <s (0x20 ^ 0x80)` for unsigned comparison. obs-text (0x80-0xFF) now accepted.

File: src/runtime/simd/avx2.cc [hunk 5]
```diff
@@ -0,0 +1,139 @@
...
+    const __m256i vcr = _mm256_set1_epi8('\r');
+    const __m256i v20 = _mm256_set1_epi8(0x20);
+    const __m256i vht = _mm256_set1_epi8(0x09);
+    const __m256i vdel = _mm256_set1_epi8(0x7F);
+
+    while (pos + 32 <= end) {
+        __m256i chunk = _mm256_loadu_si256(reinterpret_cast<const __m256i*>(buf + pos));
+        u32 cr_mask = static_cast<u32>(_mm256_movemask_epi8(_mm256_cmpeq_epi8(chunk, vcr)));
+        // bad: (< 0x20 && != HT) || == DEL
+        __m256i lt20 = _mm256_cmpgt_epi8(v20, chunk);  // v20 > chunk ≡ chunk < v20 (signed)
```
[the remark this answers is not in the document]
Comment (hurricane1026) [reply src/runtime/simd/avx2.cc]: Fixed. Same xor-0x80 unsigned comparison applied to AVX2 path.

File: testing/bench.h [hunk 6]
```diff
@@ -0,0 +1,313 @@
+#pragma once
+
+// Lightweight microbenchmark framework — zero stdlib dependency.
+// Inspired by nanobench. Uses clock_gettime(CLOCK_MONOTONIC).
+//
+// Usage:
+//   #include "rout/bench.h"
```
[the remark this answers is not in the document]
Comment (hurricane1026) [reply testing/bench.h]: Fixed. Updated comment to `#include "bench.h"`.

File: src/runtime/http_parser.cc [hunk 7]
```diff
@@ -0,0 +1,403 @@
...
+            return ParseStatus::Complete;
+        }
+    } else if (first == 't') {
+        if (name_len == 17 && str_ci_eq(name + 1, "ransfer-encoding", 16)) {
+            if (vlen >= 7 && str_ci_eq(val + vlen - 7, "chunked", 7)) {
+                req->chunked = true;
+            }
+            return ParseStatus::Complete;
+        }
+    }
```
[the remark this answers is not in the document]
Comment (hurricane1026) [reply src/runtime/http_parser.cc:243]: Fixed. TE now parsed as comma-separated token list with exact `tok_len == 7` match. "notchunked" correctly rejected — added test.

File: include/rout/runtime/http_parser.h [hunk 8]
```diff
@@ -0,0 +1,101 @@
...
+    GET = 0,
+    POST = 1,
+    PUT = 2,
+    DELETE = 3,
+    PATCH = 4,
+    HEAD = 5,
+    OPTIONS = 6,
+    CONNECT = 7,
+    TRACE = 8,
+    Unknown = 255,
```
[the remark this answers is not in the document]
Comment (hurricane1026) [reply include/rout/runtime/http_parser.h]: Fixed. Changed comment to "sequential values".

File: src/runtime/simd/neon.cc [hunk 9]
```diff
@@ -0,0 +1,156 @@
...
+    uint8x8_t sum_lo = vpadd_u8(lo, lo);  // 4 values
+    uint8x8_t sum_hi = vpadd_u8(hi, hi);
+    sum_lo = vpadd_u8(sum_lo, sum_lo);  // 2 values
+    sum_hi = vpadd_u8(sum_hi, sum_hi);
+    sum_lo = vpadd_u8(sum_lo, sum_lo);  // 1 value
+    sum_hi = vpadd_u8(sum_hi, sum_hi);
+    return static_cast<u32>(vget_lane_u8(sum_lo, 0)) |
+           (static_cast<u32>(vget_lane_u8(sum_hi, 0)) << 8);
+}
+
```
[the remark this answers is not in the document]
Comment (hurricane1026) [reply src/runtime/simd/neon.cc]: Fixed. Removed unused `neon_movemask`, `neon_first_set`, `neon_bitmask16`. Only `neon_any` remains.

File: src/runtime/http_parser.cc [hunk 10]
```diff
@@ -0,0 +1,418 @@
...
+                u32 tok_start = ti;
+                while (ti < vlen && val[ti] != ',' && val[ti] != ' ' && val[ti] != '\t') ti++;
+                u32 tok_len = ti - tok_start;
+                if (tok_len == 7 && str_ci_eq(val + tok_start, "chunked", 7)) {
+                    req->chunked = true;
+                    break;
+                }
+            }
+            return ParseStatus::Complete;
+        }
```
[the remark this answers is not in the document]
Comment (hurricane1026) [reply src/runtime/http_parser.cc:242]: Fixed. Added post-header-loop check: `if (req->chunked && req->has_content_length) return Error`. 5 tests cover this.

File: include/rout/runtime/http_parser.h [hunk 11]
```diff
@@ -0,0 +1,102 @@
...
+
+    void reset() {
+        parsed_offset = 0;
+        header_end = 0;
+    }
+
+    // Parse request-line + headers from buf[0..len).
+    // On Complete, populates `req` and sets `header_end`.
+    // On Incomplete, preserves `parsed_offset` for next call.
+    ParseStatus parse(const u8* buf, u32 len, ParsedRequest* req);
```
[the remark this answers is not in the document]
Comment (hurricane1026) [reply include/rout/runtime/http_parser.h]: Fixed. Removed `parsed_offset` field and updated all documentation.

File: tests/test_http_parser.cc [hunk 12]
```diff
@@ -0,0 +1,2434 @@
...
+    auto s = parse_one(
+        "PUT /url HTTP/1.1\r\n"
+        "Content-Length: 1\r\n"
+        "Content-Length: 2\r\n"
+        "\r\n",
+        &req,
+        &parser);
+    // Current behavior: parses successfully, takes last value
+    CHECK_EQ(static_cast<u8>(s), static_cast<u8>(ParseStatus::Complete));
+    CHECK_EQ(req.content_length, 2u);
```
[the remark this answers is not in the document]
Comment (hurricane1026) [reply tests/test_http_parser.cc]: Fixed. `DuplicateContentLength` test now expects Error. Added `DuplicateContentLengthSameValue` (expects Complete). 6 duplicate-CL tests total.

File: dev.sh [hunk 13]
```diff
@@ -75,14 +71,21 @@ coverage() {
         --object "$BUILD_DIR-cov/tests/test_network" \
         --object "$BUILD_DIR-cov/tests/test_integration" \
         --object "$BUILD_DIR-cov/tests/test_arena" \
+        --object "$BUILD_DIR-cov/tests/test_http_parser" \
```
[the remark this answers is not in the document]
Comment (hurricane1026) [reply dev.sh:74]: Fixed. Added `--object` for `test_expected` and `test_buffer` in dev.sh coverage.

File: .github/workflows/ci.yml [hunk 14]
```diff
@@ -107,6 +118,7 @@ jobs:
             --object build/tests/test_network \
             --object build/tests/test_integration \
             --object build/tests/test_arena \
+            --object build/tests/test_http_parser \
```
[the remark this answers is not in the document]
Comment (hurricane1026) [reply .github/workflows/ci.yml:123]: Fixed. Added `--object` for `test_expected` and `test_buffer` in CI coverage.

File: testing/bench.h [hunk 15]
```diff
@@ -0,0 +1,315 @@
...
+    }
+    for (int i = n - 1; i >= 0; i--) (void)::write(1, &buf[i], 1);
+}
+
+// Print fixed-point: val is in nanoseconds, print as "X.YYY ns" or "X.YYY us" or "X.YYY ms"
+inline void out_duration_ns(u64 ns) {
+    if (ns < 1000) {
+        out_u64(ns);
+        out(" ns");
+    } else if (ns < 1000000) {
```
[the remark this answers is not in the document]
Comment (hurricane1026) [reply testing/bench.h:90]: Fixed. Changed comment to "X.YY" matching the 2-decimal implementation.

File: src/runtime/http_parser.cc [hunk 16]
```diff
@@ -0,0 +1,418 @@
...
+// Core parser — single-pass, no find_header_end pre-scan
+// ============================================================================
+
+// In two-pass mode (used when parsed_offset indicates we haven't scanned yet
+// or the fast single-pass fails to find enough data), we fall back to
+// find_header_end + full re-parse.
+//
+// In single-pass mode, we parse directly and return Incomplete if we run
+// out of buffer before finding the end of headers. This eliminates the
+// pre-scan for the common case where the full request arrives in one recv.
```
[the remark this answers is not in the document]
Comment (hurricane1026) [reply src/runtime/http_parser.cc]: Fixed. Replaced two-pass comment with accurate single-pass description.

File: tests/CMakeLists.txt [hunk 17]
```diff
@@ -14,24 +18,43 @@ add_executable(test_integration test_integration.cc)
...
 add_custom_target(check
     COMMAND $<TARGET_FILE:test_network>
     COMMAND $<TARGET_FILE:test_integration>
     COMMAND $<TARGET_FILE:test_arena>
     COMMAND $<TARGET_FILE:test_expected>
+    COMMAND $<TARGET_FILE:test_http_parser>
     COMMAND $<TARGET_FILE:test_buffer>
-    DEPENDS test_network test_integration test_arena test_expected test_buffer
+    DEPENDS test_network test_integration test_arena test_expected test_http_parser test_buffer
     COMMENT "Running all tests..."
```
[the remark this answers is not in the document]
Comment (hurricane1026) [reply tests/CMakeLists.txt:59]: Fixed. Removed `test_framework` from check target (contains intentional failure).

File: src/runtime/http_parser.cc [hunk 18]
```diff
@@ -0,0 +1,418 @@
...
+// Returns quickly for non-semantic headers via first-byte + length dispatch.
+static inline ParseStatus apply_semantic_header(
+    const u8* name, u32 name_len, const u8* val, u32 vlen, ParsedRequest* req) {
+    // Dispatch on (first_byte | 0x20) and length for fast rejection
+    u8 first = name[0] | 0x20;
+
+    if (first == 'c') {
+        if (name_len == 14 && str_ci_eq(name + 1, "ontent-length", 13)) {
+            i64 cl = parse_uint(val, vlen);
+            if (UNLIKELY(cl < 0)) return ParseStatus::Error;
```
[the remark this answers is not in the document]
Comment (hurricane1026) [reply src/runtime/http_parser.cc:212]: Fixed. Added `has_content_length` flag to ParsedRequest. Duplicate CL with different values returns Error. 6 tests cover boundary cases.

File: src/runtime/http_parser.cc [hunk 19]
```diff
@@ -0,0 +1,418 @@
...
+                u32 tok_len = ti - tok_start;
+                if (tok_len == 7 && str_ci_eq(val + tok_start, "chunked", 7)) {
+                    req->chunked = true;
+                    break;
+                }
+            }
+            return ParseStatus::Complete;
+        }
+    }
+    return ParseStatus::Complete;  // not a semantic header — no-op
```
[the remark this answers is not in the document]
Comment (hurricane1026) [reply src/runtime/http_parser.cc:244]: Fixed. Post-header check rejects `chunked && has_content_length`. 5 conflict tests including CL:0+chunked.

File: test/fixtures/llhttp/lenient-header-value-relaxed.md [hunk 20]
```diff
@@ -0,0 +1,158 @@
+Relaxed header value character parsing
+=======================================
+
+Relaxed parsing mode: accepts unusual characters (like control chars)
+but still rejects specifally dangerous ones (NULL, CR, LF) that could enable
```
[the remark this answers is not in the document]
Comment (hurricane1026) [reply test/fixtures/llhttp/lenient-header-value-relaxed.md]: Fixed. Typo corrected.

File: test/fixtures/llhttp/method.md [hunk 21]
```diff
@@ -0,0 +1,524 @@
...
+off=40 header_field complete
+off=41 len=11 span[header_value]="example.com"
+off=54 header_value complete
+off=56 headers complete method=33 v=1/0 flags=0 content_length=0
+off=56 message complete
+```
+
+### OPTIONS request with RTSP
+
+NOTE: `OPTIONS` is a valid HTTP metho too.
```
[the remark this answers is not in the document]
Comment (hurricane1026) [reply test/fixtures/llhttp/method.md]: Fixed. Typo corrected.

File: src/runtime/simd/neon.cc [hunk 22]
```diff
@@ -0,0 +1,123 @@
...
+u32 scan_uri(const u8* buf, u32 pos, u32 end) {
+    const uint8x16_t vsp = vdupq_n_u8(' ');
+    const uint8x16_t v21 = vdupq_n_u8(0x21);
+    const uint8x16_t v7f = vdupq_n_u8(0x7F);
+
+    while (pos + 16 <= end) {
+        uint8x16_t chunk = vld1q_u8(buf + pos);
+        uint8x16_t sp_match = vceqq_u8(chunk, vsp);
+        uint8x16_t bad = vorrq_u8(vcltq_u8(chunk, v21), vceqq_u8(chunk, v7f));
```
[the remark this answers is not in the document]
Comment (hurricane1026) [reply src/runtime/simd/neon.cc]: Fixed. Added `vcgeq_u8(chunk, v80)` to reject bytes >= 0x80 in NEON scan_uri. 128-byte systematic test covers this.

File: src/runtime/simd/sve.cc [hunk 23]
```diff
@@ -0,0 +1,115 @@
...
+
+    while (pos < end) {
+        u32 remaining = end - pos;
+        svbool_t pred = svwhilelt_b8(static_cast<u32>(0), remaining);
+        svuint8_t chunk = svld1_u8(pred, buf + pos);
+
+        svbool_t sp_match = svcmpeq_u8(pred, chunk, ' ');
+        svbool_t lt21 = svcmplt_u8(pred, chunk, 0x21);
+        svbool_t is_del = svcmpeq_u8(pred, chunk, 0x7F);
+        svbool_t bad = svorr_b_z(pred, lt21, is_del);
```
[the remark this answers is not in the document]
Comment (hurricane1026) [reply src/runtime/simd/sve.cc]: Fixed. Added `svcmpge_u8(pred, chunk, 0x80)` to SVE scan_uri.

File: src/runtime/simd/avx512.cc [hunk 24]
```diff
@@ -0,0 +1,125 @@
...
+
+u32 scan_uri(const u8* buf, u32 pos, u32 end) {
+    const __m512i vsp = _mm512_set1_epi8(' ');
+    const __m512i v21 = _mm512_set1_epi8(0x21);
+    const __m512i v7f = _mm512_set1_epi8(0x7F);
+
+    while (pos + 64 <= end) {
+        __m512i chunk = _mm512_loadu_si512(buf + pos);
+        u64 sp_mask = _mm512_cmpeq_epi8_mask(chunk, vsp);
+        u64 bad_mask = _mm512_cmplt_epu8_mask(chunk, v21) | _mm512_cmpeq_epi8_mask(chunk, v7f);
```
[the remark this answers is not in the document]
Comment (hurricane1026) [reply src/runtime/simd/avx512.cc]: Fixed. Added `_mm512_movepi8_mask(chunk)` to extract high-bit mask in AVX-512 scan_uri.

File: tests/CMakeLists.txt [hunk 25]
```diff
@@ -14,24 +18,44 @@ add_executable(test_integration test_integration.cc)
...
 add_executable(test_buffer test_buffer.cc)
-target_include_directories(test_buffer PRIVATE ${PROJECT_SOURCE_DIR}/include)
+target_include_directories(test_buffer PRIVATE
+    ${PROJECT_SOURCE_DIR}/include
+    ${PROJECT_SOURCE_DIR}/testing
+)
 
 add_custom_target(check
+    COMMAND $<TARGET_FILE:test_framework> || true
     COMMAND $<TARGET_FILE:test_network>
```
[the remark this answers is not in the document]
Comment (hurricane1026) [reply tests/CMakeLists.txt:52]: Fixed. Removed `test_framework` from check target entirely instead of masking with `|| true`.

File: src/runtime/http_parser.cc [hunk 26]
```diff
@@ -0,0 +1,426 @@
...
+    if (UNLIKELY(len == 0)) return -1;
+    u64 val = 0;
+    for (u32 i = 0; i < len; i++) {
+        u32 d = p[i] - '0';
+        if (UNLIKELY(d > 9)) return -1;
+        u64 next = val * 10 + d;
+        if (UNLIKELY(next < val)) return -1;
+        val = next;
+    }
+    if (UNLIKELY(val > 0xFFFFFFFF)) return -1;
```
[the remark this answers is not in the document]
Comment (hurricane1026) [reply src/runtime/http_parser.cc]: Fixed. Changed to u32 with pre-multiply bound check: `val > (0xFFFFFFFFU - d) / 10`. Added 6 boundary tests (UINT32_MAX±1, 11-digit, leading zeroes).

File: src/CMakeLists.txt [hunk 27]
```diff
@@ -1,3 +1,38 @@
...
+elseif(SIMD_ARCH STREQUAL "neon")
+    set(SIMD_SOURCE runtime/simd/neon.cc)
+    set(SIMD_FLAGS "")
+elseif(SIMD_ARCH STREQUAL "sve")
+    set(SIMD_SOURCE runtime/simd/sve.cc)
+    set(SIMD_FLAGS "-march=armv8-a+sve")
+else()
+    set(SIMD_SOURCE runtime/simd/scalar.cc)
+    set(SIMD_FLAGS "")
+endif()
```
[the remark this answers is not in the document]
Comment (hurricane1026) [reply src/CMakeLists.txt:37]: Fixed. Added `message(FATAL_ERROR ...)` for unknown SIMD_ARCH values and `set_property(CACHE SIMD_ARCH PROPERTY STRINGS ...)`.

File: include/rout/runtime/http_parser.h [hunk 28]
```diff
@@ -0,0 +1,105 @@
...
```
[the remark this answers is not in the document]
Comment (hurricane1026) [reply include/rout/runtime/http_parser.h]: Fixed. Removed `parsed_offset` field — was write-only. Updated struct comment to document single-pass reparse behavior.

File: src/runtime/http_parser.cc [hunk 29]
```diff
@@ -0,0 +1,425 @@
...
+        // Find end of token
+        u32 tok_start = i;
+        while (i < vlen && val[i] != ',' && val[i] != ' ' && val[i] != '\t') i++;
+        u32 tok_len = i - tok_start;
+
+        if (tok_len == 5 && str_ci_eq(val + tok_start, "close", 5)) {
+            req->keep_alive = false;
+        } else if (tok_len == 10 && str_ci_eq(val + tok_start, "keep-alive", 10)) {
+            req->keep_alive = true;
+        }
```
[the remark this answers is not in the document]
Comment (hurricane1026) [reply src/runtime/http_parser.cc:198]: Fixed. `close` is now sticky: `match_connection` returns immediately after matching "close", preventing subsequent "keep-alive" from overriding. 2 tests verify ordering.

File: .github/workflows/ci.yml [hunk 30]
```diff
@@ -12,13 +12,23 @@ jobs:
...
+        run: |
+          wget -qO- https://apt.llvm.org/llvm-snapshot.gpg.key | sudo tee /etc/apt/trusted.gpg.d/llvm.asc
+          echo "deb http://apt.llvm.org/noble/ llvm-toolchain-noble main" | \
+            sudo tee /etc/apt/sources.list.d/llvm.list
+          sudo apt-get update
+          # LLVM snapshot publishes as clang-format-20 (dev version)
+          sudo apt-get install -y clang-format-20 || sudo apt-get install -y clang-format-19
+          # Use whichever was installed
+          CF=$(command -v clang-format-20 || command -v clang-format-19)
+          echo "CF=$CF" >> "$GITHUB_ENV"
```
[the remark this answers is not in the document]
Comment (hurricane1026) [reply .github/workflows/ci.yml]: Fixed. Updated comment to say "v20" matching the actual installed version.

File: testing/bench.h [hunk 31]
```diff
@@ -0,0 +1,315 @@
...
+    u64 epoch_iters_ = 0;  // 0 = auto (min_iters / epochs)
+    u64 bytes_per_op_ = 0;
+
+    Result results_[32];
+    u32 result_count_ = 0;
+
+    void title(const char* t) { title_ = t; }
+    void min_iterations(u64 n) { min_iters_ = n; }
+    void warmup(u64 n) { warmup_iters_ = n; }
+    void epochs(u32 n) { epochs_ = n > kMaxEpochs ? kMaxEpochs : n; }
```
[the remark this answers is not in the document]
Comment (hurricane1026) [reply testing/bench.h]: Fixed. Clamped epochs to [1, kMaxEpochs] and ensured odd for clean median.

File: include/rout/runtime/http_parser.h [hunk 32]
```diff
@@ -0,0 +1,99 @@
...
+    GET = 0,
+    POST = 1,
+    PUT = 2,
+    DELETE = 3,
+    PATCH = 4,
+    HEAD = 5,
+    OPTIONS = 6,
+    CONNECT = 7,
+    TRACE = 8,
+    Unknown = 255,
```
[the remark this answers is not in the document]
Comment (hurricane1026) [reply include/rout/runtime/http_parser.h]: Fixed. Changed to "0..8 for known methods, 255 for unknown".

File: dev.sh [hunk 33]
```diff
@@ -39,16 +40,9 @@ build() {
...
-    ninja -C "$BUILD_DIR" test_network test_integration test_arena test_expected
-    echo "=== Running mock tests ==="
-    "$BUILD_DIR/tests/test_network"
-    echo "=== Running integration tests ==="
-    "$BUILD_DIR/tests/test_integration"
-    echo "=== Running arena tests ==="
-    "$BUILD_DIR/tests/test_arena"
-    echo "=== Running expected tests ==="
-    "$BUILD_DIR/tests/test_expected"
+    build
```
[the remark this answers is not in the document]
Comment (hurricane1026) [reply dev.sh:43]: Fixed. Removed redundant `build &&` from test case arm — test() already calls build() internally.

File: src/runtime/http_parser.cc [hunk 34]
```diff
@@ -0,0 +1,423 @@
...
+    // Quick check: need at least 4 bytes to try method matching.
+    // (Shorter buffers can only be incomplete.)
+    if (UNLIKELY(len < 4)) {
+        return ParseStatus::Incomplete;
+    }
+
+    // Single-pass: parse directly without find_header_end pre-scan.
+    // We check bounds at each step and return Incomplete if needed.
+    req->reset();
+    u32 pos = 0;
```
[the remark this answers is not in the document]
Comment (hurricane1026) [reply src/runtime/http_parser.cc:267]: Fixed. Added `header_end = 0` at the start of `parse()` so stale values are never exposed on Incomplete/Error returns.

File: dev.sh [hunk 35]
```diff
@@ -75,14 +71,23 @@ coverage() {
...
 # ---- clang-tidy ----
 tidy() {
     configure
     echo "=== Running clang-tidy ==="
-    local src_cc=$(find "$PROJECT_DIR/src" -name '*.cc' | grep -v third_party)
+    # Exclude SIMD backends that require target-specific intrinsics
+    local src_cc=$(find "$PROJECT_DIR/src" -name '*.cc' \
+        ! -path '*/simd/neon.cc' \
+        ! -path '*/simd/sve.cc' \
+        ! -path '*/simd/avx2.cc' \
```
[the remark this answers is not in the document]
Comment (hurricane1026) [reply dev.sh:89]: Fixed. Excluded `sse2.cc` alongside other arch-specific SIMD files in both dev.sh and CI. Only scalar.cc and http_parser.cc are linted.

File: bench/bench_http_parser.cc [hunk 36]
```diff
@@ -0,0 +1,147 @@
...
+                rue_parser.parse(reinterpret_cast<const u8*>(payload.data), payload.len, &rue_req);
+            bench::do_not_optimize(&s);
+            bench::do_not_optimize(&rue_req);
+        });
+
+        // --- llhttp parser ---
+        llhttp_t llhttp_parser;
+
+        b.run("llhttp", [&] {
+            llhttp_init(&llhttp_parser, HTTP_REQUEST, &llhttp_settings);
```
[the remark this answers is not in the document]
Comment (hurricane1026) [reply bench/bench_http_parser.cc]: Fixed. Moved `llhttp_init()` outside the hot loop; using `llhttp_reset()` per-iteration now. Results barely changed — llhttp_init and llhttp_reset have similar cost.

File: .github/workflows/ci.yml [hunk 37]
```diff
@@ -12,13 +12,21 @@ jobs:
     runs-on: ubuntu-24.04
     steps:
       - uses: actions/checkout@v4
-      - name: Install clang-format
-        run: sudo apt-get update && sudo apt-get install -y clang-format
+      - name: Install clang-format (LLVM snapshot, v20)
+        run: |
+          wget -qO- https://apt.llvm.org/llvm-snapshot.gpg.key | sudo tee /etc/apt/trusted.gpg.d/llvm.asc
+          echo "deb http://apt.llvm.org/noble/ llvm-toolchain-noble main" | \
```
[the remark this answers is not in the document]
Comment (hurricane1026) [reply .github/workflows/ci.yml]: Fixed. Changed to `https://apt.llvm.org`.

File: testing/bench.h [hunk 38]
```diff
@@ -0,0 +1,318 @@
...
+                } else {
+                    u64 ratio = (results_[i].median_ns * 100) / base_ns;
+                    out("  ");
+                    out_u64(ratio / 100);
+                    out(".");
+                    u64 frac = ratio % 100;
+                    if (frac < 10) out("0");
+                    out_u64(frac);
+                    out("x slower than ");
+                }
```
[the remark this answers is not in the document]
Comment (hurricane1026) [reply testing/bench.h:307]: Fixed. Added `if (base_ns == 0) base_ns = 1` and `if (other_ns == 0) other_ns = 1` guards before division in `compare()`.

File: tests/CMakeLists.txt (hunk 17 above)
[the remark this answers is not in the document]
Comment (hurricane1026) [reply tests/CMakeLists.txt:59]: Not adding — `test_framework` contains an intentional failing test (`check_fail_reported` asserts `2+2==5`) that verifies the test framework reports failures correctly. It always exits 1, so adding it to `check` would break CI. This was discussed in a previous review (2969550980 / 2969589953) and the decision was to exclude it.
